{
  "manifest": {
    "name": "apollo-link-error",
    "version": "1.1.5",
    "description": "Error Apollo Link for GraphQL Network Stack",
    "author": {
      "name": "James Baxley",
      "email": "james@meteor.com"
    },
    "license": "MIT",
    "main": "./lib/bundle.umd.js",
    "module": "./lib/index.js",
    "jsnext:main": "./lib/index.js",
    "typings": "./lib/index.d.ts",
    "repository": {
      "type": "git",
      "url": "git+https://github.com/apollographql/apollo-link.git"
    },
    "bugs": {
      "url": "https://github.com/apollographql/apollo-link/issues"
    },
    "homepage": "https://github.com/apollographql/apollo-link#readme",
    "scripts": {
      "build:browser": "browserify ./lib/bundle.umd.js -o=./lib/bundle.js --i apollo-link && npm run minify:browser",
      "build": "tsc -p .",
      "bundle": "rollup -c",
      "clean": "rimraf lib/* && rimraf coverage/*",
      "coverage": "jest --coverage",
      "filesize": "npm run build && npm run build:browser",
      "lint": "tslint --type-check -p tsconfig.json -c ../../tslint.json src/*.ts",
      "minify:browser": "uglifyjs -c -m -o ./lib/bundle.min.js -- ./lib/bundle.js",
      "postbuild": "npm run bundle",
      "prebuild": "npm run clean",
      "prepublishOnly": "npm run clean && npm run build",
      "test": "jest",
      "watch": "tsc -w -p . & rollup -c -w"
    },
    "dependencies": {
      "apollo-link": "^1.2.6",
      "apollo-link-http-common": "^0.2.8"
    },
    "devDependencies": {
      "@types/graphql": "14.0.3",
      "@types/jest": "22.2.3",
      "browserify": "16.2.3",
      "graphql": "14.0.2",
      "graphql-tag": "2.10.0",
      "jest": "22.4.4",
      "rimraf": "2.6.1",
      "rollup": "0.67.4",
      "ts-jest": "22.4.6",
      "tslint": "5.11.0",
      "typescript": "3.0.3",
      "uglify-js": "3.4.9"
    },
    "jest": {
      "transform": {
        ".(ts|tsx)": "ts-jest"
      },
      "testRegex": "(/__tests__/.*|\\.(test|spec))\\.(ts|tsx|js)$",
      "moduleFileExtensions": [
        "ts",
        "tsx",
        "js",
        "json"
      ],
      "testURL": "http://localhost"
    },
    "_registry": "npm",
    "_loc": "yarn-cache/v4/npm-apollo-link-error-1.1.5-1d600dfa75c4e4bf017f50d60da7b375b53047ab/node_modules/apollo-link-error/package.json",
    "readmeFilename": "README.md",
    "readme": "---\ntitle: apollo-link-error\ndescription: Handle and inspect errors in your GraphQL network stack.\n---\n\nUse this link to do some custom logic when a GraphQL or network error happens:\n\n```js\nimport { onError } from \"apollo-link-error\";\n\nconst link = onError(({ graphQLErrors, networkError }) => {\n  if (graphQLErrors)\n    graphQLErrors.map(({ message, locations, path }) =>\n      console.log(\n        `[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`\n      )\n    );\n  if (networkError) console.log(`[Network error]: ${networkError}`);\n});\n```\n\nApollo Link is a system of modular components for GraphQL networking. [Read the docs](https://www.apollographql.com/docs/link/#usage) to learn how to use this link with libraries like Apollo Client and graphql-tools, or as a standalone client.\n\n<h2 id=\"callback\">Callback</h2>\n\nError Link takes a function that is called in the event of an error. This function is called with an object containing the following keys:\n\n* `operation`: The Operation that errored\n* `response`: The result returned from lower down in the link chain\n* `graphQLErrors`: An array of errors from the GraphQL endpoint\n* `networkError`: Any error during the link execution or server response, that wasn't delivered as part of the `errors` field in the GraphQL result\n* `forward`: A reference to the next link in the chain. Calling `return forward(operation)` in the callback will retry the request, returning a new observable for the upstream link to subscribe to.\n\nReturns: `Observable<FetchResult> | void` The error callback can optionally return an observable from calling `forward(operation)` if it wants to retry the request. It should not return anything else.\n\n<h2 id=\"error-categories\">Error categorization</h2>\n\nAn error is passed as a `networkError` if a link further down the chain called the `error` callback on the observable. In most cases, `graphQLErrors` is the `errors` field of the result from the last `next` call.\n\nA `networkError` can contain additional fields, such as a GraphQL object in the case of a [failing HTTP status code](http.html#Errors) from [`apollo-link-http`](http.html). In this situation, `graphQLErrors` is an alias for `networkError.result.errors` if the property exists.\n\n<h2 id=\"retry-request\">Retrying failed requests</h2>\n\nAn error handler might want to do more than just logging errors. You can check for a certain failure condition or error code, and retry the request if rectifying the error is possible. For example, when using some form of token based authentication, there is a need to handle re-authentication when the token expires. Here is an example of how to do this using `forward()`.\n```js\nonError(({ graphQLErrors, networkError, operation, forward }) => {\n    if (graphQLErrors) {\n      for (let err of graphQLErrors) {\n        switch (err.extensions.code) {\n          case 'UNAUTHENTICATED':\n            // error code is set to UNAUTHENTICATED\n            // when AuthenticationError thrown in resolver\n\n            // modify the operation context with a new token\n            const oldHeaders = operation.getContext().headers;\n            operation.setContext({\n              headers: {\n                ...oldHeaders,\n                authorization: getNewToken(),\n              },\n            });\n            // retry the request, returning the new observable\n            return forward(operation);\n        }\n      }\n    }\n    if (networkError) {\n      console.log(`[Network error]: ${networkError}`);\n      // if you would also like to retry automatically on\n      // network errors, we recommend that you use\n      // apollo-link-retry\n    }\n  }\n);\n```\n\nHere is a diagram of how the request flow looks like now:\n![Diagram of request flow after retrying in error links](https://i.imgur.com/ncVAdz4.png)\n\nOne caveat is that the errors from the new response from retrying the request does not get passed into the error handler again. This helps to avoid being trapped in an endless request loop when you call forward() in your error handler.\n\n\n<h2 id=\"ignoring-errors\">Ignoring errors</h2>\n\nIf you want to conditionally ignore errors, you can set `response.errors = null;` within the error handler:\n\n```js\nonError(({ response, operation }) => {\n  if (operation.operationName === \"IgnoreErrorsQuery\") {\n    response.errors = null;\n  }\n});\n```\n"
  },
  "artifacts": [],
  "remote": {
    "resolved": "https://registry.yarnpkg.com/apollo-link-error/-/apollo-link-error-1.1.5.tgz#1d600dfa75c4e4bf017f50d60da7b375b53047ab",
    "type": "tarball",
    "reference": "https://registry.yarnpkg.com/apollo-link-error/-/apollo-link-error-1.1.5.tgz",
    "hash": "1d600dfa75c4e4bf017f50d60da7b375b53047ab",
    "integrity": "sha512-gE0P711K+rI3QcTzfYhzRI9axXaiuq/emu8x8Y5NHK9jl9wxh7qmEc3ZTyGpnGFDDTXfhalmX17X5lp3RCVHDQ==",
    "registry": "npm",
    "packageName": "apollo-link-error"
  },
  "registry": "npm",
  "hash": "1d600dfa75c4e4bf017f50d60da7b375b53047ab"
}